go-controller/pkg/ovn/ovn.go

```diff
@@ -610,6 +640,112 @@ func (oc *Controller) GetServiceVIPToName(vip string, protocol kapi.Protocol) (t
 	return namespace, ok
 }
 
+// setServiceLBToACL associates an empty load balancer with its associated ACL reject rule
+func (oc *Controller) setServiceACLToLB(lb, vip, acl string) {
+	oc.serviceLBLock.Lock()
+	defer oc.serviceLBLock.Unlock()
+	if _, ok := oc.serviceLBMap[lb]; !ok {
+		oc.serviceLBMap[lb] = make(map[string]*loadBalancerConf)
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{rejectACL: acl}
+		return
+	}
+	if _, ok := oc.serviceLBMap[lb][vip]; !ok {
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{rejectACL: acl}
+		return
+	}
+	// This should not happen, the entry should be removed for the VIP when the service is deleted for an existing
+	// entry
+	if oc.serviceLBMap[lb][vip] == nil {
+		klog.Errorf("config for existing load balancer is nil. LB: %s, VIP: %s", lb, vip)
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{rejectACL: acl}
+		return
+	}
+	oc.serviceLBMap[lb][vip].rejectACL = acl
+}
+
+// setServiceEndpointsToLB associates a load balancer with endpoints
+func (oc *Controller) setServiceEndpointsToLB(lb, vip string, eps []string) {
+	oc.serviceLBLock.Lock()
+	defer oc.serviceLBLock.Unlock()
+	if _, ok := oc.serviceLBMap[lb]; !ok {
+		oc.serviceLBMap[lb] = make(map[string]*loadBalancerConf)
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{endpoints: eps}
+		return
+	}
+	if _, ok := oc.serviceLBMap[lb][vip]; !ok {
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{endpoints: eps}
+		return
+	}
+	// This should not happen, the entry should be removed for the VIP when the service is deleted
+	if oc.serviceLBMap[lb][vip] == nil {
+		klog.Errorf("config for existing load balancer is nil. LB: %s, VIP: %s", lb, vip)
+		oc.serviceLBMap[lb][vip] = &loadBalancerConf{endpoints: eps}
+		return
+	}
+	oc.serviceLBMap[lb][vip].endpoints = eps
+}
+
+func (oc *Controller) getServiceLBACL(lb, vip string) string {
+	return oc.getServiceLBConfig(lb, vip).rejectACL
+}
+
+func (oc *Controller) serviceLBHasEndpoints(lb, vip string) bool {
+	return len(oc.getServiceLBConfig(lb, vip).endpoints) > 0
+}
+
+// getServiceLBConfig retrieves the associated Config for a given load balancer
+func (oc *Controller) getServiceLBConfig(lb, vip string) loadBalancerConf {
+	oc.serviceLBLock.Lock()
+	defer oc.serviceLBLock.Unlock()
+	conf, ok := oc.serviceLBMap[lb][vip]
+	if !ok {
+		conf = &loadBalancerConf{}
+	}
+	return *conf
```

Comment:
@trozet it's an odd pattern to return the dereferenced object which will result in a struct copy to the caller. Instead, what about replacing all 3 functions with:

```
// getServiceLBInfo returns the reject ACL and whether the number of endpoints for the service is greater than zero
func (oc *Controller) getServiceLBInfo(lb, vip string) (string, bool) {
	oc.serviceLBLock.Lock()
	defer oc.serviceLBLock.Unlock()
	conf, ok := oc.serviceLBMap[lb][vip]
	if !ok {
		conf = &loadBalancerConf{}
	}
	return conf.rejectACL, len(conf.endpoints) > 0
}
```

and then just ignore the return value you don't care about, eg:

```
if _, hasEPs := ovn.serviceLBInfo(loadBalancer, vip); hasEPs {
```

Reply:
I don't think it is that odd. It's just returning the value instead of a pointer, but I'll change it as you suggest.